src/engine/renderer/Material.h

```diff
@@ -0,0 +1,202 @@
+
+===========================================================================
+*/
+// Material.h
+
+#ifndef MATERIAL_H
+#define MATERIAL_H
+
+#include <vector>
+
+#include "gl_shader.h"
+#include "tr_local.h"
+
+static constexpr uint MAX_DRAWCOMMAND_TEXTURES = 64;
+
+struct DrawCommand {
+	GLIndirectBuffer::GLIndirectCommand cmd;
+	uint materialsSSBOOffset = 0;
+	uint textureCount = 0;
+	Texture* textures[MAX_DRAWCOMMAND_TEXTURES];
+
+	DrawCommand() {
+	}
+
+	DrawCommand( const DrawCommand& other ) {
+		cmd = other.cmd;
+		materialsSSBOOffset = other.materialsSSBOOffset;
+		textureCount = other.textureCount;
+		memcpy( textures, other.textures, textureCount * sizeof( Texture* ) );
+	}
+};
+
+struct Material {
+	uint materialsSSBOOffset = 0;
+	uint staticMaterialsSSBOOffset = 0;
+	uint dynamicMaterialsSSBOOffset = 0;
+	uint totalDrawSurfCount = 0;
+	uint totalStaticDrawSurfCount = 0;
+	uint totalDynamicDrawSurfCount = 0;
+	uint currentDrawSurfCount = 0;
+	uint currentStaticDrawSurfCount = 0;
+	uint currentDynamicDrawSurfCount = 0;
+
+	uint staticCommandOffset = 0;
+
+	uint id = 0;
+	bool useSync = false;
+	uint syncMaterial = 0; // Must not be drawn before the material with this id
+
+	uint32_t stateBits = 0;
+	stageType_t stageType;
+	GLuint program = 0;
+	GLShader* shader;
+
+	int deformIndex;
+	bool vertexAnimation;
+	bool tcGenEnvironment;
+	bool tcGen_Lightmap;
+	bool hasDepthFade;
+	bool vboVertexSprite;
+	bool alphaTest;
+
+	bool bspSurface;
+	bool enableDeluxeMapping;
+	bool enableGridLighting;
+	bool enableGridDeluxeMapping;
+	bool hasHeightMapInNormalMap;
+	bool enableReliefMapping;
+	bool enableNormalMapping;
+	bool enablePhysicalMapping;
+
+	cullType_t cullType;
+
+	bool usePolygonOffset = false;
+
+	VBO_t* vbo;
+	IBO_t* ibo;
+
+	std::vector<drawSurf_t*> drawSurfs;
+	std::vector<DrawCommand> drawCommands;
+	bool texturesResident = false;
+	std::vector<Texture*> textures;
+
+	bool operator==( const Material& other ) {
+		return program == other.program && stateBits == other.stateBits && vbo == other.vbo && ibo == other.ibo
+			&& cullType == other.cullType && usePolygonOffset == other.usePolygonOffset;
+	}
+
+	void AddTexture( Texture* texture ) {
+		if ( !texture->hasBindlessHandle ) {
+			texture->GenBindlessHandle();
+		}
+
+		if ( std::find( textures.begin(), textures.end(), texture ) == textures.end() ) {
+			textures.emplace_back( texture );
+		}
+	}
+};
+
+struct drawSurfBoundingSphere {
+	vec3_t origin;
+	float radius;
+
+	uint drawSurfID;
+};
+
+class MaterialSystem {
+	public:
+	bool generatedWorldCommandBuffer = false;
+	bool skipDrawCommands;
+	bool skipSurface;
+	bool generatingWorldCommandBuffer = false;
+	vec3_t worldViewBounds[2] = {};
+
+	std::vector<drawSurf_t*> portalSurfacesTmp;
+	std::vector<drawSurf_t> portalSurfaces;
+	std::vector<drawSurfBoundingSphere> portalBounds;
+	std::vector<shader_t*> skyShaders;
+
+	std::vector<Material*> renderedMaterials;
+
+	struct MaterialPack {
```

Comment:
It's just an abstraction to match a range of materials with the 3 different calls to `RB_RenderDrawSurfaces()` with 3 different `shaderSort_t` ranges in `RB_RenderView()`. The 4th one that uses a different surface filter (`DRAWSURFACES_NEAR_ENTITIES`) is ignored because it's never used for BSP surfaces.

Reply:
I would appreciate this comment in code.

Reply:
Will add in #1137.